### PR TITLE
[LLVM][rtsan] Add sanitize_realtime attribute for the realtime sanitizer

### PR DESCRIPTION
**Pull request by cjappl**
Add a new "sanitize_realtime" attribute, which will correspond to the sanitize_realtime function effect in clang. This is used in the realtime sanitizer transform

Please see the [reviewer support document](https://github.com/realtime-sanitizer/radsan/blob/doc/review-support/doc/review.md) for what our next steps are. The original discourse thread can be found [here](https://discourse.llvm.org/t/rfc-nolock-and-noalloc-attributes/76837)

### CODE REVIEW

**Comment by cjappl** [comment]
CC for review @vitalybuka @MaskRay @Sirraide @AaronBallman  (maybe you can help suggest if someone is more appropriate in LLVM, this is my first PR in this area)

Co-author @davidtrevelyan

File: llvm/include/llvm/Transforms/Instrumentation/RealtimeSanitizer.h
Comment (cjappl) [inline llvm/include/llvm/Transforms/Instrumentation/RealtimeSanitizer.h:20]: This is intentionally empty for now, as we anticipate having to add options very soon

File: llvm/include/llvm/Transforms/Instrumentation/RealtimeSanitizerOptions.h
Comment (cjappl) [inline llvm/include/llvm/Transforms/Instrumentation/RealtimeSanitizerOptions.h:15]: Similarly to the above comment, we anticipate needing these options in the near future, so decided to introduce this with the boilerplate. Happy to remove this file if anyone wants

File: llvm/test/Instrumentation/RealtimeSanitizer/rtsan.ll
Comment (cjappl) [inline llvm/test/Instrumentation/RealtimeSanitizer/rtsan.ll:34]: One question we had: We want to be able to insert this before EVERY function return. However I was not able to generate an example that ever had multiple return statements in llvm-ir.

Is it possible to have multiple return statements in IR?
If so, what is a C/C++ example that induces this?

Adding a test to ensure it is added before every return statement would be a great improvement.

**Review by vitalybuka — APPROVED** [review]
LGTM but better to move transforms and CODE_OWNERS into separate patches

**Comment by cjappl** [comment]
I have updated the review, splitting it into three:

1. The codeowners review https://github.com/llvm/llvm-project/pull/101231
2. This review introducing the attribute
3. The review introducing the transform, which directly depends on this review https://github.com/llvm/llvm-project/pull/101232

**Review by nikic — CHANGES_REQUESTED** [review]
Missing LangRef specification.

**Comment by nikic** [comment]
Can't say for sure without seeing the spec, but this is almost certainly the wrong name for the attribute. `nonblocking` would imply that "blocking" (for some rigorous definition thereof) is UB and LLVM can optimize with that knowledge. What you probably want is a `sanitize_realtime` attribute or something.

**Comment by cjappl** [comment]
> Can't say for sure without seeing the spec, but this is almost certainly the wrong name for the attribute. `nonblocking` would imply that "blocking" (for some rigorous definition thereof) is UB and LLVM can optimize with that knowledge. What you probably want is a `sanitize_realtime` attribute or something.

This name was chosen to match the name of the function effect it keys off of. This sanitizer works hand in hand with clang "function effects" that look like `[[clang::nonblocking]]` and `[[clang::blocking]]`
https://discourse.llvm.org/t/rfc-nolock-and-noalloc-attributes/76837

A "reviewers guide" can be found here, describing how the sanitizer works:
https://github.com/realtime-sanitizer/radsan/blob/doc/review-support/doc/review.md

We will defer to people more experienced in LLVM what the appropriate choice is here!

**Comment by cjappl** [comment]
> Missing LangRef specification.

Fixed in latest version! Thanks for catching

File: llvm/test/Instrumentation/RealtimeSanitizer/rtsan.ll
Comment (vitalybuka) [reply llvm/test/Instrumentation/RealtimeSanitizer/rtsan.ll]: > Is it possible to have multiple return statements in IR?
Yes

> If so, what is a C/C++ example that induces this?
Not sure.

> Adding a test to ensure it is added before every return statement would be a great improvement.
You can just write that IR whithout c++

**Review by vitalybuka — APPROVED** [review]

File: llvm/test/Instrumentation/RealtimeSanitizer/rtsan.ll
Comment (cjappl) [reply llvm/test/Instrumentation/RealtimeSanitizer/rtsan.ll]: I will investigate and add this to the other review if I crack it!! Thanks for the hint.

This review is now just the attribute, I'll move these comments over there.

**Comment by cjappl** [comment]
@nikic based on your suggestions, I decided to update this to the attribute `sanitize_realtime`. Let me know if this works for you or you want to see other changes.

**Review by nikic — APPROVED** [review]
LGTM

**Comment by cjappl** [comment]
@nikic or @vitalybuka could we get a merge on this when you get a chance?

Thanks for everyone's review and discussion!

**Pull request merged**
